lib/jsonrpc-adapter.js

```diff
@@ -148,33 +148,41 @@ JsonRpcAdapter.prototype.createHandler = function() {
 
     methods.forEach(function(method) {
       // Wrap the method so that it will keep its own receiver - the shared class
-      var fn = function() {
-        var args = arguments;
-        if (method.isStatic) {
-          method.getFunction().apply(method.ctor, args);
-        } else {
-          method.sharedCtor.invoke(method, function(err, instance) {
-            method.getFunction().apply(instance, args);
-          });
-        }
-      };
-
-      /*
-      I had to this because jayson uses fn.toString to map the parameters,
-      and since you wrap the method, the jsonrpc server cannot read them.
-      I solved this overwriting the toString method on the wrapped function,
-      creating this fake string that has the paramenters names in it.
-      */
+      var argsNames = '';
       if (method.accepts) {
-        var argsNames = method.accepts.map(function(item) {
+        argsNames = method.accepts.map(function(item) {
           return item.arg;
         });
-
-        fn.toString = function() {
-          return 'function (' + argsNames.concat('callback').join(',') + '){}';
-        };
+        argsNames = argsNames.join(',');
+      } else {
+        var m = method.getFunction();
+        if (m.length > 1) {
+          // The method has more args than cb
+          // Build dummy param names
+          var names = [];
+          for (var i = 0; i < m.length - 1; i++) {
+            names.push('param' + i);
+          }
+          argsNames = names.join(',');
+        }
+      }
+      argsNames = argsNames ? argsNames + ',cb' : 'cb';
+      var funcBody =
+        'return function(' + argsNames + ') {\n' +
+        '  var args = Array.prototype.slice.call(arguments);\n' +
+        '  if (method.isStatic) {\n' +
+        '    method.getFunction().apply(method.ctor, args);\n' +
+        '  } else {\n' +
+        '    method.sharedCtor.invoke(method, function(err, instance) {\n' +
+        '      method.getFunction().apply(instance, args);\n' +
+        '    });\n' +
+        '  }\n' +
+        '};';
```

Comment:
Can we use `Function.toString` instead of writing the code in a string?

```js
// mockup
var wrapper = function wrapper() {
  var args = //..
  // ... etc
};
var funcBody = wrapper.toString()
  .replace(/^function wrapper()/, 'function(' + argsNames + '));
var fn = new Function('method', funcBody)(method);
```